docs/ai/conceptual/understanding-tokens.md

```diff
@@ -0,0 +1,121 @@
+- _Pros_:
+  - Allows the model to handle a wider range of inputs, such as unknown words, typos, or complex syntax
+  - Might allow the vocabulary size to be reduced, requiring less memory resources
+- _Cons_:
+  - A given text is broken into more tokens, requiring additional computational resources while processing
+  - Given a fixed token limit, the maximum size of the model's input and output is smaller
+
+**Larger tokens (word tokenization)**
+
+- _Pros_:
+  - A given text is broken into fewer tokens, requiring less computational resources while processing
+  - Given the same token limit, the maximum size of the model's input and output is larger
+- _Cons_:
+  - Might cause an increased vocabulary size, requiring more memory resources
+  - Can limit the models ability to handle unknown words, typos, or complex syntax
+
+## How tokens are used
+
+While training, after a text has been tokenized, each unique token is assigned an ID.
+
+Consider the following sentence:
+
+> I heard a dog bark loudly at a cat
+
+After tokenization, the following IDs could be assigned:
+
+- I (1)
+- heard (2)
+- a (3)
+- dog (4)
+- bark (5)
+- loudly (6)
+- at (7)
+- a (the "a" token is already assigned an ID of 3)
+- cat (8)
+
+By assigning IDs, text can be represented as a sequence of token IDs. The previous sentence would be represented as [1, 2, 3, 4, 5, 6, 7, 3, 8]. The sentence "I heard a cat" would be represented as [1, 2, 3, 8].
+
+As training continues, any new tokens in the training text are added to the vocabulary and assigned an ID. For example:
+
+- meow (9)
+- run (10)
+
+The semantic relationships between the tokens can be analyzed using these token ID sequences. These relationships are represented as multi-valued numeric vectors, known as [_embeddings_](understanding-embeddings.md). Each token is assigned an embedding based on how commonly it's used together with or in similar contexts to the other tokens.
+
+During output generation, the model predicts a vector value for the next token in the sequence. This vector value is then used to select the next token from the model's vocabulary. In practice, multiple vectors are calculated using various elements of the preceding tokens' embeddings. The model then evaluates all potential tokens from these vectors and selects the most probable one to continue the sequence.
+
+Output generation is an iterative operation. The predicted token is appended to the sequence so far and used as the input for the next iteration, building the final output one token at a time.
+
+### Token limits
+
+LLMs have limits on the maximum number of tokens that can be used as input context or generated as output. However, the limits for input context and output generation can differ.
+
+For example, GPT-4 supports up to 128,000 tokens of input context but only 4,096 tokens of output.
+
+Together a model's token limit and tokenization method determine the maximum length of text that can be provided as input or generated as output.
```

Comment:
This is a nice insight - perhaps use the running example to provide a contrasting pair of examples (such as character-based and word-based)?

Reply:
Added an example of how many input and output tokens would be used/available with word or character tokenzation.